LowStar libraries for `vector` and the `regional` typeclass

joonwonc opened the pull request:
LowStar libraries for `vector` and the `regional` typeclass:
- LowStar.Vector.fst: `vector`, a resizable `LowStar.Buffer` where a struct has its current size and capacity
- LowStar.Regional.fst: the `regional` typeclass for fine-grained region control
- LowStar.RVector.fst: `vector` as a `regional` typeclass instance
- LowStar.Regional.Instances.fst: some useful typeclass instances

joonwonc commented:
@msprotz and @tahina-pro : I would appreciate if you can take a look at the code and give some feedback!

msprotz commented on ulib/LowStar.Regional.fst:
```diff
@@ -0,0 +1,79 @@
+    // An initialization operation. We might have a multiple ways of
+    // initializing a given target type `a`; then multiple typeclass instances
+    // should be defined, and each of them can be used properly.
+    r_init: (r:erid ->
```
Hello Joonwon, Tahina and I are starting a quick review of this... sorry for the delay.
```suggestion
    r_alloc: (r:erid ->
```

msprotz replied:
The rationale being that we generally use `init` for things that have been allocated (e.g. un-initialized buffers) already but for which we haven't written the initial state.

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+
+/// Capacity
+
+unfold val size_of: #a:Type -> vec:vector a -> Tot uint32_t
```
Why all the `unfold`s? I would suggest `inline_for_extraction` if you're hoping to see these helpers inlined at call-site

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+
+/// Abstract vector type
+
+noeq type vector_str a =
```
You might want to add `inline_for_extraction` on the type too, if you're hoping to see the projectors for the type inlined, too.

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+		  loc_disjoint (loc_vector vec) dloc /\
+		  modifies dloc h0 h1))
+	(ensures (as_seq h0 vec == as_seq h1 vec))
+	[SMTPat (live h0 vec); 
```
This pattern most likely won't work, owing to `live` defined as `unfold`

msprotz replied:
the `unfold`s above are likely causing a bunch of issues, since `loc_vector` is also defined as unfold; Tahina and I recommend removing the `unfold`s above, and leaving `inline_for_extraction` for the small helpers you hope to see disappear when extracting; if there was a deeper reason for using `unfold`, let us know!

(the general rule of thumb is that a definition marked as unfold is inlined before going to SMT, meaning that it will not be able to be matched against)

msprotz replied:
Tahina says it's a good practice to write `live h1 vec` in your post-condition because it's almost always going to be useful and will save z3 some search effort.

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+
+/// Construction
+
+val create_empty: 
```
this would be a good candidate for `inline_for_extraction`

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+let create #a len v =
+  create_rid len v HH.root
+
+val create_reserve:
```
for this function, and possibly others around here, I would recommend writing a brief documentation that states in a single sentence what this function does compared to the other `create*` functions

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+let create_reserve #a len ia rid =
+  Vec 0ul len (B.malloc rid ia len)
+
+val create_by_buffer:
```
Tahina recommends naming these `create*` based on the fact that `alloc`/`free` oftentimes go hand-in-hand

msprotz replied:
(just like in the `Buffer` library)

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+    (B.upd vs sz v;
+    Vec (sz + 1ul) cap vs)
+
+val flush:
```
this one also could use a brief description

msprotz commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,558 @@
+    (decreases (B.length buf))
+let rec fold_left_buffer #a #b len buf f ib =
+  if len = 0ul then ib
+  else (fold_left_buffer (len - 1ul) (B.sub buf 1ul (len - 1ul)) 
```
This one could conceivably use the `C.Loops` module for a more efficient iteration, but that can come in a later pull request

msprotz replied:
(especially since we need to move C.Loops to LowStar.Loops first)

msprotz commented on ulib/LowStar.RVector.fst:
```diff
@@ -0,0 +1,1060 @@
+  let r1 = HST.new_region r0 in
+  let hh1 = HST.get () in
+  B.modifies_none_modifies hh0 hh1;
+  r1
```
Tahina thinks that `new_region_modifies` from `LowStar.Monotonic.Buffer` should trigger. Do you know why you've had to define a separate helper?

joonwonc commented:
Thanks a lot for the review!
I actually needed some changes in the `regional` typeclass and `rvector` in the meantime.
Let me try to apply both the changes and the review, and to commit a revision as soon as possible.

msprotz commented:
The build is failing because of:
```
Error: Unbound type constructor FStar_Set.equal
```
I seem to recall this issue because you hit it earlier and @aseemr gave you a workaround. Does that ring a bell?

@tahina-pro can you do a pass on the updated changes!

Thanks for the fast turnaround time @joonwonc

msprotz commented:
I'll let @aseemr tell us whether the workaround is acceptable for now or not -- I'd say yes.

aseemr commented:
Re. `FStar_Set.ml:equal`, I suggested @joonwonc to add it, but with a comment that F* should not extract it. Once CMI related extraction changes are stable, we can revisit this.

joonwonc commented:
I pushed the change to `FStar_Set.ml` with the comment.

msprotz commented:
Perfect. Waiting for CI then I'll merge.

msprotz commented:
@nikswamy  said he'd take care of the final merge... deferring to him. Thanks!

nikswamy commented on ulib/LowStar.Regional.fst:
```diff
@@ -0,0 +1,79 @@
+module HH = FStar.Monotonic.HyperHeap
+module HS = FStar.HyperStack
+module HST = FStar.HyperStack.ST
+module MHS = FStar.Monotonic.HyperStack
```
Why distinguish `MHS` from `HS`? The latter includes the former

joonwonc replied:
Yes, no specific reason for this one; do you want me to fix it through all related files?

joonwonc replied:
Fixed in the below commit!

joonwonc commented:
I pushed one more fix in `LowStar.RVector` that caused memory leak.

joonwonc commented:
Hello @nikswamy :

Thanks for the review, I really appreciate it.
Here are my answers for your questions:

> I don't understand how the representation chosen in LowStar.Vector.vector_str allows you to implement that This looks to me just like a thin wrapper around buffer allowing you to grow a buffer of a larger initial capacity than needed, allowing it to grow (up to the initial capactity) into the unused suffix.

Some operations of `vector` deal with reallocation. For instance, when `insert` is called, it tries to resize the capacity (by reallocating a buffer) when it's filled. This mechanism is abstracted so users don't need to worry about resizing while calling any operations.

> Can we have an interface for LowStar.Vector and LowStar.RVector?

I tried this, but found that a lot of predicate *definitions* are dependent to each other so they should be revealed in interface. If you think it's ok then I can work on it.

> Why the `dummy` element in the regional type class?

Sometimes we want to allocate (initialize) a data where an initial element is stateless (e.g., initializing a vector of buffers, filled with `null`). In order to have a general typeclass instance (e.g., `regional a -> regional (vector a)`) I had to have a general notion of `dummy` in `regional`.

> Why the separate `r_alloc_p: (a -> GTot Type0)`: Can't an instance just include whatever this predicate is meant to capture in the main `r_inv` invariant?

`r_alloc_p` is not an invariant, but a property that should hold for initial data. For instance, for `vector`s, `r_alloc_p` could say the length of the vector is zero. In my opinion, this property cannot be captured by `r_inv`.

> FStar.HyperStack.ST.erid is already defined: no need to duplicate it in Regional

Fixed.

> Should `region_of` in regional return an `erid`?

Currently it returns `rid` (`region_of: (a -> GTot HH.rid)`). I think `rid` is enough since it's mostly used to check the disjointness of `rid`s.

> Let's not open FStar.All, FStar.Classical etc. when they are not actually used

Fixed.

> Independently, a procedural issue: I'd like each file to be begin with the following copyright (especially since the work was done while at MSR) and license.

Fixed.

Let me know if you have more questions or suggestions!

nikswamy commented on ulib/LowStar.Vector.fst:
```diff
@@ -0,0 +1,687 @@
+      			  (loc_vector nvec)) h0 h1 /\
+      size_of nvec = size_of vec + 1ul /\
+      get h1 nvec (size_of vec) == v /\
+      S.equal (as_seq h1 nvec) (S.snoc (as_seq h0 vec) v)))
```
@joonwonc Sorry for the slow reviews on this ...

I see now that `insert` resizes the vector, but it does by reallocating and passing back a resized vector, rather than mutating the original.

This is fine, but it is not the design goal described to me by @ad-l @protz and @fournet 

Is this expected by all?

joonwonc replied:
I got your point, two things to mention here:

1. If you meant to maintain the original vector **struct** but to reallocate a **buffer**, this is possible by passing a vector struct **pointer** to the function and changing field values appropriately. As I remember, at the time I built this library we didn't have a way to use a field access with a struct pointer (`->`), so it currently returns a new vector struct. But as far as I know we can use `->` now so I could change the interface.

2. If you meant to maintain the original **buffer** pointer, we need a way to change the length of an allocated buffer. As far as I know we don't have such mechanism in C. So we just need to allocate a new buffer with a new capacity and to copy the content. This mechanism is also used by C++ std::vector, and for example, because of that, whenever we add en element to a vector, all iterators for the vector are invalidated (https://en.cppreference.com/w/cpp/container/vector/push_back).

msprotz replied:
Passing a pointer to a struct then relying on KreMLin's pretty-printing to compile:

```
p.(0ul) <- { p.(0ul) with underlying_buffer with the_newly_allocated_buffer }
```

as:
```
p->underlying_buffer = the_newly_allocated_buffer
```
is always feasible on top of the current API, I believe (possibly with suitable inlining). However, this would be quite a burden to clients of this API, having to manage the outer pointer.

Since the struct is (I believe) three words in C, I would assume that making this function a struct-returning one, rather than an in-place mutating one, won't have too much of a performance impact.

I remember having similar discussions with @tahina-pro and also advocating for the struct-returning style, in order to limit the burden on clients.

So, in short, the current style is fine for me, and we can always define the more C-idiomatic style on top of it, should we want to.

nikswamy replied:
I don't see why the client should have to explicitly manipulate a pointer to a struct with reliance on KreMLin for better code etc.

I was expecting the library to internally represent a vector as a pointer to a buffer, its current size and max capacity ... with buffer like operations (index, update) etc. lifted to this representation, and insert possibly resizing the buffer in case we're out of capacity. This should be completely transparent to clients.

aseemr commented:
Hi @joonwonc. First off, thanks for the library and your patience with the review!

My main comment is about the use of `FStar.Monotonic.HyperHeap`. It is an internal module that clients should not use directly. `FStar.HyperStack` includes it, and so, all its symbols should be available under the `HS.` namespace. We left it there for backward compatibility but F* could give a warning if the clients open it directly. A better documentation could also help, currently there is just a comment in `FStar.Monotonic.HyperHeap`.

Could you make this change in the meantime.

aseemr commented:
Could we also want a monotonic vector in the future? If so, this library should use `LowStar.Monotonic.Buffer` and should have a preorder.

We should merge it as is, but something to keep in mind. Also, changing it in the future would be easy enough.

aseemr commented:
Oh another minor nit: the change in `FStar_Set.ml` is already in master (since @R1kM needed it for his development).

joonwonc commented:
Hello @aseemr:

- I removed all use of `FStar.Monotonic.HyperHeap`.
- I'm not sure whether we will need monotonic vector for now, but as you said we can add the feature after merging, whenever needed.
- About `FStar_Set.ml`: I would prefer to resolve this possible conflict when merging, by choosing the file in `master`.

Thanks a lot for the review! Let me know if you have any more suggestions or questions.

nikswamy commented:
I just merged this to master ... thanks Joonwon! Especially for your patience with the long discussion here.

The merge was overdue and other projects have been waiting for this code. But, there are still some aspects of it that can be improved.

You'll notice that I added a high-level description of the module to LowStar.Regional and LowStar.Vector.

It would be great to have comments of a similar flavor for LowStar.RVector and LowStar.Regional.Instances.

Another question is why the `copyable` class is in LowStar.RVector: doesn't it deserve its own file, like Regional? Also, can we have a more descriptive name instead of `RVector`? e.g., LowStar.Regional.Vector

Finally, do we really need a LowStar.Regional.Instances file? That's not particularly modular, e.g., as we add more instances. Is it feasible to give the instance for vector in LowStar.RVector/LowStar.Regional.Vector.

We could also have a LowStar.Regional.Buffer for the buffer instance.

nikswamy commented:
As a bonus, it would be great to have another type, e.g., `LowStar.Resizable.Vector`, implemented as pointer to a LowStar.Vector, where resizing can be done imperatively rather than via struct passing (as discussed above and on slack).

joonwonc commented:
@nikswamy : thanks for merging this pull request! Now I can work on using this library directly in my Merkle tree implementation in `hacl-star`.

I would be happy to improve the library with additional pull requests, so yes, I'd appreciate if you can make the pull request.
